server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -635,16 +632,16 @@ private ThreadContextStruct putResponse(
 
             final Map<String, Set<String>> newResponseHeaders;
             final Set<String> existingValues = responseHeaders.get(key);
+            if (existingValues != null && existingValues.contains(uniqueValue.apply(value))) {
+                return this;
+            }
+            newResponseHeaders = new HashMap<>(responseHeaders);
             if (existingValues != null) {
-                if (existingValues.contains(uniqueValue.apply(value))) {
-                    return this;
-                }
                 // preserve insertion order
-                final Set<String> newValues = Stream.concat(existingValues.stream(), Stream.of(value)).collect(LINKED_HASH_SET_COLLECTOR);
-                newResponseHeaders = new HashMap<>(responseHeaders);
+                final Set<String> newValues = new LinkedHashSet<>(existingValues);
+                newValues.add(value);
                 newResponseHeaders.put(key, Collections.unmodifiableSet(newValues));
             } else {
-                newResponseHeaders = new HashMap<>(responseHeaders);
                 newResponseHeaders.put(key, Collections.singleton(value));
             }
```

Comment:
Is it worth to simplify this if-else with a Map#merge?

Reply:
Not really I think because we already pay for the get that gives us `existingValues` so it's really just extra indirection.